src/undocommand.js

```diff
@@ -0,0 +1,195 @@
+/**
+ * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+import Command from '../command/command.js';
+
+/**
+ * Undo command stores batches in itself and is able to and apply reverted versions of them on the document.
+ *
+ * @memberOf undo
+ */
+export default class UndoCommand extends Command {
+	constructor( editor ) {
+		super( editor );
+
+		/**
+		 * Items that are pairs of batches which are saved by the command and model selection state at the moment of saving the batch.
+		 *
+		 * @private
+		 * @member {Array.<engine.treeModel.Batch>} undo.UndoCommand#_batchStack
+		 */
+		this._batchStack = [];
+	}
+
+	/**
+	 * Stores a batch in the command. Stored batches can be then reverted.
+	 *
+	 * @param {engine.treeModel.Batch} batch Batch to add.
+	 */
+	addBatch( batch ) {
+		const selection = {
+			ranges: Array.from( this.editor.document.selection.getRanges() ),
+			isBackward: this.editor.document.selection.isBackward
+		};
+
+		this._batchStack.push( { batch, selection } );
+	}
+
+	/**
+	 * Removes all batches from the stack.
+	 */
+	clearStack() {
+		this._batchStack = [];
+	}
+
+	_checkEnabled() {
+		return this._batchStack.length > 0;
+	}
+
+	/**
+	 * Executes the command: reverts a {@link engine.treeModel.Batch batch} added to the command's stack,
+	 * applies it on the document and removes the batch from the stack.
+	 *
+	 * @private
+	 * @fires undo.undoCommand#event:revert
+	 * @param {Number} [batchIndex] If set, batch under the given index on the stack will be reverted and removed.
+	 * If not set, or invalid, the last added batch will be reverted and removed.
+	 */
+	_doExecute( batchIndex ) {
```

Comment:
We haven't thought about this but we need to document somewhere (public) the command's params. Or we can assume that the doc builder will somehow handle that.

Second thing is that I agree with @pjasiun that `batchIndex` is not a right param since you cannot retrieve it from anywhere. You don't even in what it is an index. Can't we pass there the batch itself and find the index?